Adding a maximum for query limit

br41nslug opened the pull request:
Implements environment variable to set a maximum value for `limit`. If this variable is set then `-1` will be substituted for the max and values over will throw an invalidQuery exception.
This implementation only applies to the API query, services are still able to fetch all using -1 independent of the setting.

Fixes ENG-487

br41nslug committed:
Adding a maximum for query limit

br41nslug committed:
fixed bad env check

br41nslug committed:
using the getEnv function to make it better testable

br41nslug committed:
updated sanitize-query tests

br41nslug committed:
added tests for the validate query function

rijkvanzanten commented on api/src/utils/sanitize-query.ts:
```diff
@@ -1,18 +1,28 @@
+				query.limit = limit;
+			}
 		}
+	} else if (hasMaxLimit) {
+		// fall back to the max limit when omitted
+		query.limit = Number(env.MAX_QUERY_LIMIT);
```
There should be a difference between default limit and max limit. Lets have env vars for `LIMIT_DEFAULT` and `LIMIT_MAX`

br41nslug replied:
Very right, we keeping that at the current hard-coded `100` default?

rijkvanzanten replied:
Yeah 👍🏻

br41nslug committed:
Added DEFAULT_QUERY_LIMIT as fallback instead

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

azrikahar commented on api/src/database/run-ast.ts:
```diff
@@ -469,7 +469,7 @@ function mergeWithParentItems(
 			if (nestedNode.query.page && nestedNode.query.page > 1) {
 				parentItem[nestedNode.fieldKey] = parentItem[nestedNode.fieldKey].slice(
-					(nestedNode.query.limit ?? 100) * (nestedNode.query.page - 1)
+					(nestedNode.query.limit ?? Number(env.DEFAULT_QUERY_LIMIT)) * (nestedNode.query.page - 1)
```
Should we add `DEFAULT_QUERY_LIMIT` and `MAX_QUERY_LIMIT` as numbers into the env typemap to reduce the need to cast them as `Number()`? 

https://github.com/directus/directus/blob/1bfd46044928a8192fbef7c1a507899b31642abf/api/src/env.ts#L274-L292

I assume this will help https://github.com/directus/directus/issues/9521#issuecomment-962502798 a little when that issue is being worked 👍

---

though understandably we're technically also not enforcing type for `STORAGE_LOCATIONS` for example, since it seems like we're doing `toArray(env.STORAGE_LOCATIONS)` whenever it's being used:

https://github.com/directus/directus/blob/1bfd46044928a8192fbef7c1a507899b31642abf/api/src/controllers/files.ts#L48

so perhaps those needs to be addressed in a separate PR.

rijkvanzanten replied:
No, lets keep casting it where it's being used. I want to get rid of the top-level casting (https://github.com/directus/directus/issues/9521)

azrikahar replied:
> No, lets keep casting it where it's being used. I want to get rid of the top-level casting

Just to confirm I understand this correctly, this would mean we wouldn't have the list of hardcoded environment variables to type cast anymore? (as apposed to what was suggested in https://github.com/directus/directus/issues/9521#issuecomment-962502798, but understandably it can be outdated 👍)

Or did you perhaps meant getting rid of _auto_ type casting, not getting rid of those existing enforced type casting?

rijkvanzanten replied:
Good question.. I think it might be easier to maintain over time if we expect every environment variable to be a strong coming out of env.ts. That also matches how env vars work in general 🤔

licitdev replied:
> easier to maintain over time if we expect every environment variable to be a strong coming out of env.ts

Strongly agree with this. 👍 Should we proceed with this refactor for the env vars in this PR?

rijkvanzanten replied:
Lets do it for the new env vars added in this PR, but lets keep the refactoring of the other env vars for another PR 👍🏻

rijkvanzanten commented on api/src/utils/sanitize-query.ts:
```diff
@@ -1,18 +1,27 @@
 		if (typeof limit === 'number') {
-			query.limit = limit;
+			if (limit === -1 && hasMaxLimit) {
+				query.limit = Number(env.MAX_QUERY_LIMIT);
```
Does this still allow for disabling the limit altogether? I think we have to add an additional `&& env.MAX_QUERY_LIMIT !== Infinity` 🤔

br41nslug replied:
Setting infinity from a string environment variable can be tricky. The current disabled state is `MAX_QUERY_LIMIT=-1` (which is also the default as that is current behavior) in which case `hasMaxLimit` will be `false` stopping the limit from being altered by this new logic.

rijkvanzanten replied:
> Setting infinity from a string environment variable can be tricky.

Why? We're expecting the string to represent a number, and `Number('Infinity')` correctly parses to JS `Infinity`:

<img width="161" alt="CleanShot 2023-02-10 at 13 59 43@2x" src="https://user-images.githubusercontent.com/9141017/218175455-6dd9404d-e25a-4a04-a07d-f83e2b50bb75.png">

br41nslug replied:
So you can, but why introduce a new value for that when we've always used -1 for that functionality

rijkvanzanten replied:
The env var `MAX_QUERY_LIMIT` controls what `-1` means in the context of the limit parameter, setting `MAX_QUERY_LIMIT` to `-1` means "-1 means -1", whereas `MAX_QUERY_LIMIT=Infinity` means "-1 means Infinity". The latter is a little clearer to me 🙂

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
make the linter happy

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

licitdev committed:
Merge branch 'main' into feat/maximum-limit

licitdev commented on api/src/utils/validate-query.ts:
```diff
@@ -11,7 +11,10 @@ const querySchema = Joi.object({
 	sort: Joi.array().items(Joi.string()),
 	filter: Joi.object({}).unknown(),
-	limit: Joi.number().integer().min(-1),
+	limit:
+		'MAX_QUERY_LIMIT' in env && env.MAX_QUERY_LIMIT !== -1
+			? Joi.number().integer().min(-1).max(env.MAX_QUERY_LIMIT) // min should be 0
+			: Joi.number().integer().min(-1),
```
I think we should also validate the requested limit with the `DEFAULT_QUERY_LIMIT`, so that it errors when the explicitly specified limit within a query exceeds. It's an invalid limit. 🤔 

```suggestion
	limit:
		'MAX_QUERY_LIMIT' in env && env.MAX_QUERY_LIMIT !== -1
			? Joi.number().integer().min(-1).max(env.MAX_QUERY_LIMIT) // min should be 0
			: Joi.number().integer().min(-1).max(env.DEFAULT_QUERY_LIMIT),
```

licitdev replied:
Incorrect interpretation of `DEFAULT_QUERY_LIMIT` being a limit itself, it acts as the limit only when a limit is not provided in a query.

rijkvanzanten commented:
One thing that we didn't talk about in the context of this change—which is also not something I'm sure is required for this PR to land—is adding some sort of `meta` information about the "next" page. Right now, `-1` will act like "max", but that also means that there's no easy way to retrieve all items. If we add a `next` flag to `meta`, you could do something like (pseudo code):

```js
const allData = [];

async function get(page) {
  const res = fetch(page);
  const { meta: { next }, data } = await res.json();

  allData.push(...data);

  if (next) {
    await get(page);
  }
}

await get('/items/example?limit=-1&page=1');
```

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
accept infinity as max_query_limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

azrikahar commented on api/src/utils/validate-query.ts:
```diff
@@ -11,7 +11,10 @@ const querySchema = Joi.object({
-	limit: Joi.number().integer().min(-1),
+	limit:
+		'MAX_QUERY_LIMIT' in env && env.MAX_QUERY_LIMIT !== -1
+			? Joi.number().integer().min(-1).max(env.MAX_QUERY_LIMIT) // min should be 0
```
I might have lost some context, but just to double check, is the comment "min should be 0" still accurate here?

br41nslug replied:
I think that comment is outdated as long as `limit=-1` still functions as the max when used in a filter.

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

azrikahar reviewed (commented):
This just occured to me, but currently the calendar layout has a set limit of `10000` (due to #13724):

https://github.com/directus/directus/blob/baf778fddb8a1c0778542ad338b0f6c85a48a7d2/app/src/layouts/calendar/index.ts#L105-L115

so when particular setup sets `MAX_QUERY_LIMIT` to lower than `10000`, it would unfortunately error. I'm not sure how can we best resolve this 🤔 a reflexive thought would be making the max query limit available in an endpoint so the Data Studio can know what is the max limit and sort of set that number into pinia store? Does feel somewhat excessive... so I do wonder if there's any better alternatives.

### Before setting MAX_QUERY_LIMIT

https://user-images.githubusercontent.com/42867097/224373347-1044f610-bd08-496b-8ff4-ccc400ebb1e2.mp4

### After setting MAX_QUERY_LIMIT to 9999

https://user-images.githubusercontent.com/42867097/224373261-35f8d1d2-6103-4c28-a3d1-9a363af56c32.mp4

---

On that note, I think a similar issue may also occur with layouts that have paginations, since currently Per Page is up to `500` and `1000`:

![](https://user-images.githubusercontent.com/42867097/224372431-01fd2071-e752-4532-82f8-e2ee51f6718c.png)

which I assume will also get affected when a setup sets a max limit that is lower than any of those predefined amounts.

rijkvanzanten commented:
@azrikahar I think we'll have to build a little "download all" util function that has a "limit" flag that we can use for use cases like this. That way it can loop over multiple requests of max limit `-1` until we hit the dataset we're looking fo r

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
fixed merge conflicts

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
added missing import

br41nslug committed:
moved docs PR into main repo

br41nslug committed:
added missing import

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

rijkvanzanten committed:
Merge branch 'main' into feat/maximum-limit

rijkvanzanten reviewed (dismissed):
2 things, otherwise LGTM!

1. Add in the app-communication piece to @azrikahar's point. We should return the max/default query information in the `/server/info` endpoint for authenticated users, and make sure the app only presents page-size options that fit in that limit
2. Lets rename `MAX_QUERY_LIMIT` and `DEFAULT_QUERY_LIMIT` to 🔁 be `QUERY_LIMIT_DEFAULT` and `QUERY_LIMIT_MAX` 🙂

br41nslug committed:
renamed vars

br41nslug committed:
basic page size limiting

rijkvanzanten committed:
Merge branch 'main' into feat/maximum-limit

rijkvanzanten committed:
Merge branch 'main' into feat/maximum-limit

rijkvanzanten commented on app/src/layouts/map/map.vue:
```diff
@@ -216,7 +195,33 @@ export default defineComponent({
+				},
+			];
+			if (info.queryLimit) {
+				return sizes.filter(({ value }) => value < info.queryLimit!.max);
```
```suggestion
				return sizes.filter(({ value }) => value <= info.queryLimit!.max);
```

rijkvanzanten commented on app/src/layouts/tabular/tabular.vue:
```diff
@@ -256,6 +257,15 @@ useShortcut(
+const { info } = useServerStore();
+
+const pageSizes = computed(() => {
+	const sizes = ['25', '50', '100', '250', '500', ' 1000'];
```
```suggestion
	const sizes = ['25', '50', '100', '250', '500', '1000'];
```

rijkvanzanten commented on app/src/layouts/tabular/tabular.vue:
```diff
@@ -256,6 +257,15 @@ useShortcut(
+const pageSizes = computed(() => {
+	const sizes = ['25', '50', '100', '250', '500', ' 1000'];
+	if (info.queryLimit) {
+		return sizes.filter((size) => Number(size) < info.queryLimit!.max);
```
```suggestion
		return sizes.filter((size) => Number(size) <= info.queryLimit!.max);
```

rijkvanzanten committed:
Update app/src/layouts/cards/cards.vue

rijkvanzanten committed:
Update app/src/layouts/map/map.vue

rijkvanzanten committed:
Update app/src/layouts/tabular/tabular.vue

rijkvanzanten committed:
Update app/src/layouts/tabular/tabular.vue

rijkvanzanten committed:
Create paginate util

rijkvanzanten commented:
I've created a new paginate utility that we should be able to use for the cases where the app currently uses `limit: -1` 🙂 

It should be a doable replacement, like:

```js
// Before
await api.get('/items/example', { params: { limit: -1, sort: 'id' }});

// After
await paginate('/items/example', { params: { sort: 'id' }}, info.queryLimit!.max);
```

https://github.com/directus/directus/pull/17309/files#diff-411728814ea78772169101c7585660ea2502d0ffe99fa4c5176cad9393b2fce2

Wanna give that a shot @azrikahar  / @br41nslug? 🙂

rijkvanzanten commented:
Lets also make sure we use the configured default as the default for pagination in the app 🙂

br41nslug committed:
extracted page size logic to composable and implemented default usage in app

br41nslug committed:
updated paginate utility

br41nslug committed:
using the new fetch-all utility

br41nslug committed:
removed limit: -1 for api calls that do not support it

br41nslug commented:
Renamed the `paginate` utility to `fetchAll` and implemented it across the app. A couple of api calls to core collections like `/collections`, `/fields`, `/relations` don't abide by query parameters and return all regardless so these cannot be used with this utility because it will loop infinitely requesting new pages.

br41nslug committed:
fix linter

br41nslug committed:
removed console log

br41nslug committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
making the linter happy

rijkvanzanten committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
Create loud-flowers-yawn.md

paescuj committed:
Merge branch 'main' into feat/maximum-limit

paescuj committed:
Merge branch 'main' into feat/maximum-limit

paescuj committed:
Merge branch 'main' into feat/maximum-limit

paescuj commented on api/src/utils/sanitize-query.test.ts:
```diff
@@ -27,6 +27,30 @@ describe('limit', () => {
 	});
 });
 
+describe('max limit', () => {
```
Should we also test the case when `QUERY_LIMIT_MAX` is undefined / -1?

paescuj committed:
Add tests for no / unlimited limit

paescuj committed:
Cover all cases for max limit in sanitizeQuery

paescuj committed:
Get rid of non-null assertion

paescuj committed:
Merge branch 'main' into feat/maximum-limit

br41nslug committed:
added test for composable

br41nslug committed:
Merge branch 'feat/maximum-limit' of https://github.com/directus/directus into feat/maximum-limit

br41nslug commented on app/src/composables/use-page-size.ts:
```diff
@@ -0,0 +1,32 @@
+	} = useServerStore();
+
+	const pageSizes = computed<T[]>(() => {
+		if (queryLimit === undefined) return availableSizes.map(mapCallback);
```
yep, added the check and some tests for this composable 😄

paescuj commented on api/src/utils/sanitize-query.ts:
```diff
@@ -1,18 +1,29 @@
+			query.limit = limit === -1 && hasMaxLimit ? Number(env['QUERY_LIMIT_MAX']) : limit;
 		}
+	} else if (hasMaxLimit) {
+		query.limit = Math.min(Number(env['QUERY_LIMIT_DEFAULT']), Number(env['QUERY_LIMIT_MAX']));
```
Is it intended that the `QUERY_LIMIT_DEFAULT` only applies if `QUERY_LIMIT_MAX` is defined?

br41nslug replied:
`QUERY_LIMIT_DEFAULT` is a fallback and gets applied here https://github.com/directus/directus/blob/e3d355f35327b62cb4bfbebf05962fde01465f47/api/src/database/run-ast.ts#L475-L490

If `limit` is not defined on the query then `sanitizeQuery` should not add it.

paescuj replied:
Thank you! That makes sense!
Could we maybe make the doc a little clearer on this, for example like this:
`QUERY_LIMIT_DEFAULT` - The default query limit used **when `QUERY_LIMIT_MAX` is set and no limit is** defined in the API request.

br41nslug replied:
It does get applied in the quoted code, independent of `QUERY_LIMIT_MAX`

paescuj replied:
Isn't that just for `o2m`?

br41nslug replied:
Hmmm looks like you're right, that seems to have been an oversight on my part 🤔 more hardcoded `100` limits should be using this environment var like https://github.com/directus/directus/blob/e3d355f35327b62cb4bfbebf05962fde01465f47/api/src/database/run-ast.ts#L253

br41nslug replied:
That latest change should properly set the `QUERY_LIMIT_DEFAULT` value if no limit is supplied

paescuj committed:
Update var names in documentation

br41nslug committed:
replace hardcoded limit default with env var

paescuj commented on app/src/composables/use-page-size.ts:
```diff
@@ -0,0 +1,45 @@
+		return sizes.map(mapCallback);
+	});
+
+	let initialSize = fallbackSize;
+
+	if (queryLimit !== undefined) {
+		initialSize = Math.min(parseLimit(queryLimit.default), parseLimit(queryLimit.max));
+	}
```
Could we perhaps simply use the first value of `pageSizes`as the initial size?
Right now, with the default value of `100` for `QUERY_LIMIT_DEFAULT` it feels like a too high default/start value in the app. Also, the value from `QUERY_LIMIT_DEFAULT` doesn't necessarily exist in the list of page sizes,  which could be confusing.

br41nslug replied:
Hmm interesting, `100` is already the default on the API side however this gets overwritten on the App side to be `25` in most cases using the `fallbackSize`. Should we perhaps not take the `default` into account in the App? 🤔

paescuj replied:
> Should we perhaps not take the default into account in the App? 

I think so 🤔

br41nslug committed:
update default page size behavior

br41nslug committed:
fixed test

paescuj reviewed (approved):
Looks & works great 🚀

azrikahar committed:
Merge branch 'main' into feat/maximum-limit

azrikahar committed:
update server info api reference

azrikahar committed:
include queryLimit in graphql server info

azrikahar committed:
use configured max limit in calendar layout

azrikahar committed:
use configured max limit in kanban layout

azrikahar committed:
account for max query limit in use-revisions
